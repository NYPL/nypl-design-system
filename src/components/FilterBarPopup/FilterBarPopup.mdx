import { ArgTypes, Canvas, Description, Meta, Source } from "@storybook/blocks";
import { changelogData } from "./filterBarPopupChangelogData";
import ComponentChangelogTable from "../../utils/ComponentChangelogTable";
import * as FilterBarPopupStories from "./FilterBarPopup.stories";
import Link from "../Link/Link";

<Meta of={FilterBarPopupStories} />

# FilterBarPopup

| Component Version | DS Version   |
| ----------------- | ------------ |
| Added             | `Prerelease` |
| Latest            | `Prerelease` |

## Table of Contents

- {<Link href="#overview" target="_self">Overview</Link>}
- {<Link href="#component-props" target="_self">Component Props</Link>}
- {<Link href="#accessibility" target="_self">Accessibility</Link>}
- {<Link href="#filterbarpopup-heading" target="_self">FilterBarPopup Heading</Link>}
- {<Link href="#controlling-state-using-selecteditems-onclear-and-onsubmit-props" target="_self">Controlling State Using selectedItems, onClear, and onSubmit Props</Link>}
- {<Link href="#totalresults-example" target="_self">totalResults example</Link>}
- {<Link href="#usefilterbarpopup-hook" target="_self">useFilterBarPopup hook</Link>}
- {<Link href="#changelog" target="_self">Changelog</Link>}

## Overview

<Description of={FilterBarPopupStories} />

`FilterBarPopup` is a wrapper component for filter components and individual form
input components. The DS considers the following components as filter components:
`CheckboxGroup`, `MultiSelectGroup`,`RadioGroup`, `Checkbox`, `DatePicker`,
`MultiSelect`, `Radio`, `Select`, `Slider`, `TextInput`, and `Toggle`.

The wrapped components/component groups can be displayed in a column layout.

It is recommended to use real-time filtering in conjunction with the `totalResults`
prop to update the `Show X results` button with X being the total number of results
associated to the results list.

If the consuming app is not using the above, then `FilterBarPopup` will render the
`Show results` button and can render an _optional_ `Clear all filters`
button which is controlled by the `onClear` prop. The `Show results` button will
call the `onSubmit` function if passed and close the modal.

Related component: [FilterBarInline](../?path=/docs/components-form-elements-FilterBarInline--docs)

## Component Props

<Canvas of={FilterBarPopupStories.WithControls} />

<ArgTypes of={FilterBarPopupStories.WithControls} />

## Accessibility

The `FilterBarPopup` component is a wrapper component for components with filter
functionality. The accessibility of `FilterBarPopup` depends on the passed child
components accessibility. See the accessibility details of valid FilterBarPopup child
components:

- [CheckboxGroup](../?path=/docs/components-form-elements-checkboxgroup--docs#accessibility)
- [MultiSelectGroup](../?path=/docs/components-form-elements-multiselectgroup--docs#accessibility)
- [RadioGroup](../?path=/docs/components-form-elements-radiogroup--docs#accessibility)
- [Checkbox](../?path=/docs/components-form-elements-checkbox--docs#accessibility)
- [DatePicker](../?path=/docs/components-form-elements-datepicker--docs#accessibility)
- [MultiSelect](../?path=/docs/components-form-elements-multiselect--docs#accessibility)
- [Radio](../?path=/docs/components-form-elements-radio--docs#accessibility)
- [Select](../?path=/docs/components-form-elements-select--docs#accessibility)
- [Slider](../?path=/docs/components-form-elements-slider--docs#accessibility)
- [TextInput](../?path=/docs/components-form-elements-textinput--docs#accessibility)
- [Toggle](../?path=/docs/components-form-elements-toggle--docs#accessibility)

## FilterBarPopup Heading

The `FilterBarPopup` component will always render a heading using the default
"Filters" label value. The default label value can be overridden using the
`heading` prop. By default, when a string is passed to the `heading` prop, the
`FilterBarPopup` component will render an `h2` element sized as an `heading5`
heading. If this causes accessibility issues in your app, you can set a
custom heading by passing in a DS `Heading` component.

In the following examples, the `FilterBarPopup` component will render custom `h3`
and `h5` elements. Note that no matter what heading level is set, the size will
always be set to `heading5` internally by the `FilterBarPopup` component.

<Source
  code={`
const customH3 = <Heading level="h3">Custom H3 Heading</Heading>;
const customH5 = <Heading level="h5">Custom H5 Heading</Heading>;
<FilterBarPopup heading={customH3} {...props} />
<FilterBarPopup heading={customH5} {...props} />
`}
  language="tsx"
/>

<Canvas of={FilterBarPopupStories.FilterBarPopupHeading} />

## Passing Children

<Canvas of={FilterBarPopupStories.MultipleChildren} />

In order to control the styling and props passed to children form elements,
the `FilterBarPopup` component accepts `CheckboxGroup`, `MultiSelectGroup`,
`RadioGroup`, `Checkbox`, `DatePicker`, `MultiSelect`, `Radio`, `Select`,
`Slider`, `TextInput`, and `Toggle` children through its `renderChildren` render
prop function.

This callback prop function returns an object argument with three properties:
`isBlockElement`, `layout`, and `width`.

- The `isBlockElement` argument value should be passed to the `isBlockElement`
  prop of `MultiSelect` children.
- The `layout` argument value should be passed to the `layout` prop of
  `CheckboxGroup`, `MultiSelectGroup`, and `RadioGroup` children.
- The `width` argument value should be passed to the `multiSelectWidth` prop of
  `MultiSelectGroup` children and the `width` prop of `MultiSelect` children.

Here is an example of the shape of `FilterBarPopup`:

<Source
  code={`
<FilterBarPopup
  id="..."
  heading="..."
  renderChildren={({ isBlockElement, layout, width }) => {
    return (...)
  }}
/>
`}
  language="jsx"
/>

## Controlling State Using selectedItems, onClear, and onSubmit Props

### selectedItems

The `FilterBarPopup` component does not store its state internally. It expects
a `selectedItems` prop whose value is an array of selected values of the filter
children.

To control each of the child form elements, you must pass the appropriate
`onChange` and `value` props to them. Find documentation on how to do this on each
individual component page.

`useMultiSelect` is an optional but recommended hook for cases where managing
the state of the component in the consuming app is less of a concern and general
ease of use is prefered. Find the full documentation under
[useMultiSelect](../?path=/docs/hooks-usemultiselect--docs)

<Source
  code={`
const { onChange, onMixedStateChange, selectedItems, onClear, onClearAll } = useMultiSelect();
const [selectedCheckbox, setSelectedCheckbox] = useState([]);
const [textValue, setTextValue] = useState("");
const [selectedFilterItems, setSelectedFilterItems] = useState([
  selectedItems,
  selectedCheckbox,
  textValue,
]);
useEffect(() => {
  setSelectedFilterItems([selectedItems, selectedCheckbox, textValue]);
}, [selectedItems, selectedCheckbox, textValue]);
`}
  language="jsx"
/>

### onClear

The `Clear all filters` button will only be visible when the `onClear` prop is passed.

The `onClear` handler will be called when the `Clear all filters` button is clicked.

It is expected that the `onClear` handler will clear all of the current values from the
child form elements and restore the associated search results list to an unfiltered
state. Consuming apps are responsible for handling the logic and subsequent UI
updates expected when the `onClear` handler is called.

<Source
  code={`
const onClearFilterBar = () => {
  onClearAll();
  setSelectedCheckbox([]);
  setTextValue("");
};`}
  language="jsx"
/>

### onSubmit

The `onSubmit` handler will be called when the `Show results` button is clicked.

It is expected that the `onSubmit` handler will read the current values from the
child form elements (the filters) and update the associated search results list
based on those values. Consuming apps are responsible for handling the logic
and subsequent UI updates expected when the `onSubmit` handler is called.

To have better readability of the code it is recommended to separate nested render
prop functions:

<Source
  code={`
// function for renderMultiSelect prop of MultiSelectGroup 
const renderMultiSelect = ({ isBlockElement, multiSelectWidth }) => {
  return (
    shortMultiSelectItems &&
    shortMultiSelectItems.map((multiSelect) => (
      <MultiSelect
        buttonText={multiSelect.name}
        key={multiSelect.id}
        id={multiSelect.id}
        items={multiSelect.items}
        selectedItems={selectedItems}
        isBlockElement={isBlockElement}
        onChange={(e) => onChange(e.target.id, multiSelect.id)}
        onMixedStateChange={(e) => {
          return onMixedStateChange(e.target.id, multiSelect.id, multiSelect.items);
        }}
        onClear={() => onClear(multiSelect.id)}
        width={multiSelectWidth}
      />
    )) 
  ); 
};`}
  language="jsx"
/>

<Source
  code={`
// function for renderChildren prop of FilterBar 
const renderFilterComponents = ({ layout, width }) => {
  return (
    <>
      <MultiSelectGroup
        id="multiselect-group"
        labelText="MultiSelect Group"
        layout={layout}
        multiSelectWidth={width}
        renderMultiSelect={renderMultiSelect}
      />
      <CheckboxGroup
        id="checkbox-example"
        labelText="Checkbox Group"
        name="checkboxExample"
        layout={layout}
        value={selectedCheckbox}
        onChange={(e) => {
          setSelectedCheckbox(e);
        }}
      >
        <Checkbox id="checkbox-1" value="1" labelText="Checkbox 1" />
        <Checkbox id="checkbox-2" value="2" labelText="Checkbox 2" />
      </CheckboxGroup>
      <TextInput
        id="textinput-example"
        isClearable
        isClearableCallback={() => setTextValue("")}
        labelText="What is your favorite color?"
        onChange={(e) => setTextValue(e.target.value)}
        placeholder="i.e. blue, green, etc."
        value={textValue}
      />
    </>
  ); 
};
`}
  language="jsx"
/>

<Source
  code={`
return (
  <FilterBarPopup
    id={args.id}
    heading={args.heading}
    layout={args.layout}
    onClear={onClearFilterBar}
    onSubmit={() => console.log(selectedFilterItems)}
    selectedItems={selectedFilterItems}
    renderChildren={renderFilterComponents}
  />
); `}
  language="jsx"
/>

## totalResults example

The `totalResults` prop should be used if the consuming app is utilizing real-time
filtering. If the `totalResults` prop is passed, its value will be used as part of
the label text for the `Show results` button (ex. "Show 72 results" where 72 is the
value of the totalResults prop). This value should be updated as filter options are
selected or changed.

The example below is not fully functioning and is meant to show how the `totalResults`
prop is shown to users in the `Show results` button.

<Canvas of={FilterBarPopupStories.TotalResultsExample} />

<Source
  code={`
const [selectedCheckbox, setSelectedCheckbox] = useState([]);
const [textValue, setTextValue] = useState("");
const [totalResults, setTotalResults] = useState(100);
const [selectedFilterItems, setSelectedFilterItems] = useState([
  selectedCheckbox,
  textValue,
]);

const updateTotalResults = () => {
// function to update totalResults value on filter change
};

...

<FilterBarPopup
  id="..."
  heading="..."
  onClear={...}
  renderChildren={({ isBlockElement, layout, width }) => {
    return (...)
  }}
  totalResults={totalResults}
/>
`}
  language="jsx"
/>

## useFilterBarPopup hook

The `useFilterBarPopup` hook provides programmatic control for the `FilterBarPopup` component
and allows consuming apps to open the modal overlay using a custom button or methods other than
a standard button click. Instead of importing the `FilterBarPopup` component directly, use
the `useFilterBarPopup` hook to get the `FilterBarPopup` component and helper functions.

<Source
  code={`
import { useFilterBarPopup } from "@nypl/design-system-react-components";
`}
  language="jsx"
/>

This hook will return an object with the `FilterBarPopup` component, a boolean value, and two functions.

<Source
  code={`
const { FilterBarPopup, isOpen, onClose, onOpen } = useFilterBarPopup();
`}
  language="jsx"
/>

The `FilterBarPopup` component is the same as the one imported directly, the `isOpen` value, and `onClose`
and `onOpen` functions are exposed and available to the consuming application. The only function that will
be used directly is the `onOpen` function. Pass `isOpen` and `onClose` to the `FilterBarPopup`.

<Source
  code={`
const { onOpen, isOpen, onClose, FilterBarPopup } = useFilterBarPopup();
// ...
<FilterBarPopup
  isOpen={isOpen}
  onClose={onClose}
  onOpen={onOpen}
  {...otherProps}
/>
`}
  language="jsx"
/>

The `onOpen` function can be used to programmatically open the `FilterBarPopup` component through another
element or behavior in the consuming app. See the example below that uses a custom `Button` to open the `Modal`.
The existing button that is rendered by the `FilterBarPopup` component will still work as expected.

<Source
  code={`
import { useFilterBarPopup } from "@nypl/design-system-react-components";
// ...
const MyComponent = () => {
  const { onOpen, isOpen, onClose, FilterBarPopup } = useFilterBarPopup();
  // ...
  return (
    <>
      <Button id="open-filter-bar-popup" onClick={onOpen}>
        Manually Open FilterBarPopup
      </Button>
      <FilterBarPopup
        isOpen={isOpen}
        onClose={onClose}
        onOpen={onOpen}
        {...otherProps}
      />
    </>
  );
}; `}
  language="jsx"
/>

## Changelog

<ComponentChangelogTable changelogData={changelogData} />
