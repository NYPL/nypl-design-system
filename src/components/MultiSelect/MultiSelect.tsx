import { chakra } from "@chakra-ui/react";
import React, { forwardRef } from "react";

import MultiSelectListBox from "./MultiSelectListbox";
import MultiSelectDialog from "./MultiSelectDialog";

export interface MultiSelectItem {
  id: string;
  name: string;
  children?: MultiSelectItem[];
}

export const multiSelectWidthsArray = [
  "default",
  "fitContent",
  "full",
] as const;
export type MultiSelectWidths = typeof multiSelectWidthsArray[number];

export interface SelectedItems {
  [name: string]: { items: string[] };
}
interface MultiSelectCommonProps {
  /** The id of the MultiSelect. */
  id: string;
  /** Set the default open or closed state of the Multiselect. */
  isDefaultOpen?: boolean;
  /** Boolean value used to control how the MultiSelect component will render within the page and interact with other DOM elements.
   * The default value is false. */
  isBlockElement?: boolean;
  /** The items to be rendered in the Multiselect as checkbox options. */
  items: MultiSelectItem[];
  /** The label text rendered within the MultiSelect. */
  labelText: string;
  /** The action to perform for clear/reset button of MultiSelect. */
  onClear?: () => void;
  /** The type of MultiSelect that will be rendered. */
  type: "listbox" | "dialog";
  /** The selected items state (items that were checked by user). */
  selectedItems: SelectedItems;
  /** Value used to set the width for the MultiSelect component. */
  width?: MultiSelectWidths;
}
type ListboxOnChange = (selectedItem: MultiSelectItem, id: string) => void;
type DialogOnChange = (event: React.ChangeEvent<HTMLInputElement>) => void;

type MultiSelectVariantsProps =
  | {
      type: "listbox";
      onApply?: never;
      /** The action to perform for downshift's onSelectedItemChange function. */
      onChange: ListboxOnChange;
      // These are props that are never allowed on the listbox type.
      onMixedStateChange?: never;
    }
  | {
      type: "dialog";
      /** The action to perform for save/apply button of multiselect. */
      onApply: () => void;
      /** The action to perform on the checkbox's onChange function.  */
      onChange: DialogOnChange;
      /** The action to perform for a mixed state checkbox (parent checkbox). */
      onMixedStateChange?: DialogOnChange;
    };

export type MultiSelectProps = MultiSelectCommonProps &
  MultiSelectVariantsProps;

/**
 * The `MultiSelect` component is a form input element that presents a list
 * of `Checkbox` components from which a user can make one or multiple
 * selections. Two variants of the MultiSelect component are offered, each with
 * slightly different functionality and requirements.  Because of these
 * differences, the two variants are broken out in separate stories below.
 */
export const MultiSelect = chakra(
  forwardRef<HTMLDivElement, React.PropsWithChildren<MultiSelectProps>>(
    (props, ref?) => {
      const {
        id,
        isBlockElement = false,
        isDefaultOpen = false,
        items,
        labelText,
        onApply,
        onChange,
        onClear,
        onMixedStateChange,
        selectedItems,
        type,
        width = "default",
        ...rest
      } = props;

      const commonProps = {
        id,
        isBlockElement,
        isDefaultOpen,
        items,
        labelText,
        onClear,
        selectedItems,
        type,
        width,
      };

      if (type === "listbox") {
        const listboxOnChange = onChange as ListboxOnChange;

        return (
          <MultiSelectListBox
            {...commonProps}
            ref={ref}
            onChange={listboxOnChange}
            {...rest}
          />
        );
      }

      if (type === "dialog") {
        const dialogOnChange = onChange as DialogOnChange;

        return (
          <MultiSelectDialog
            {...commonProps}
            onChange={dialogOnChange}
            onMixedStateChange={onMixedStateChange}
            onApply={onApply}
            ref={ref}
            {...rest}
          />
        );
      }

      return null;
    }
  ),
  // Pass all custom props to Chakra and override, for width prop.
  { shouldForwardProp: () => true }
);

export default MultiSelect;
