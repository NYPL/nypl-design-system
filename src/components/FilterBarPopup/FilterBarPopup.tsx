import {
  Box,
  chakra,
  ChakraComponent,
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalFooter,
  ModalHeader,
  ModalOverlay,
  Stack,
  useDisclosure,
  useMultiStyleConfig,
} from "@chakra-ui/react";
import React, { forwardRef, useRef } from "react";
import Button from "../Button/Button";
import ButtonGroup from "../ButtonGroup/ButtonGroup";
import Heading, { HeadingSizes } from "../Heading/Heading";
import { LayoutTypes } from "../../helpers/types";
import { MultiSelectWidths } from "../MultiSelect/MultiSelect";
import { FilterBarItemsType } from "../FilterBarInline/FilterBarInline";

export interface FilterBarPopupProps {
  /** The className of the FilterBarInline. */
  className?: string;
  /** Optional string value used to set the text for a `Heading` component, or
   * a DS Heading component that can be passed in.
   */
  heading?: string | JSX.Element;
  /** ID that other components can cross reference for accessibility purposes. */
  id?: string;
  /** Only used for internal purposes. */
  isOpen?: boolean;
  /** Function for the global `Clear Filters` button. If passed the `Clear Filters` button will render. */
  onClear?: () => void;
  /** Only used for internal purposes. */
  onClose?: any;
  /** Only used for internal purposes. */
  onOpen?: any;
  /** Function for the global `Apply Filters` button. If passed the `Apply Filters` button will render. */
  onSubmit?: () => void;
  /** Render prop function used to render child components with updated props. */
  renderChildren: ({
    isBlockElement,
    layout,
    width,
  }: {
    isBlockElement?: boolean;
    layout?: LayoutTypes;
    width?: MultiSelectWidths;
  }) => React.ReactNode;
  /** The selected items state (items that were checked by user). */
  selectedItems?: FilterBarItemsType;
  /** A number that represents the total number of results in the associated search results list. */
  totalResults?: number;
}

export const FilterBarPopup: ChakraComponent<
  React.ForwardRefExoticComponent<
    React.PropsWithChildren<FilterBarPopupProps> &
      React.RefAttributes<HTMLDivElement>
  >,
  FilterBarPopupProps
> = chakra(
  forwardRef<HTMLDivElement, React.PropsWithChildren<FilterBarPopupProps>>(
    (props, ref?) => {
      const {
        children,
        id,
        isOpen,
        heading = "Filters",
        onClear,
        onSubmit,
        onOpen,
        onClose,
        renderChildren,
        selectedItems,
        totalResults,
        ...rest
      } = props;

      const styles = useMultiStyleConfig("FilterBarPopup", {
        width: "full",
      });

      const generalHeadingProps = {
        size: "heading5" as HeadingSizes,
        noSpace: true,
      };
      // If `heading is a string, then we want the default heading,
      // otherwise, use whatever the user passed in.
      const finalHeading = heading ? (
        typeof heading === "string" ? (
          <Heading level="h2" text={heading} {...generalHeadingProps} />
        ) : (
          React.cloneElement(heading, generalHeadingProps)
        )
      ) : null;

      // Chakra's hook to control Modal's actions.
      const disclosure = useDisclosure();
      const finalIsOpen = isOpen ? isOpen : disclosure.isOpen;
      const finalOnOpen = onOpen ? onOpen : disclosure.onOpen;
      const finalOnClose = onClose ? onClose : disclosure.onClose;

      const onSubmitAndClose = () => {
        onSubmit();
        finalOnClose();
      };

      const showResultsButtonRef: React.RefObject<HTMLButtonElement> =
        useRef<HTMLButtonElement>();

      const onClearAndFocus = () => {
        onClear();

        setTimeout(() => {
          showResultsButtonRef.current?.focus();
        }, 1);
      };

      return (
        <Box id={`filter-bar-${id}`} ref={ref} {...rest}>
          <Button
            id={`filter-bar-${id}-show-filters`}
            buttonType="secondary"
            onClick={finalOnOpen}
            width={{ base: "100%", md: "fit-content" }}
          >
            {`Show filters`}
          </Button>
          <Modal
            isOpen={finalIsOpen}
            onClose={finalOnClose}
            size={{ base: "full", md: "md" }}
            scrollBehavior="inside"
          >
            <ModalOverlay />
            <ModalContent>
              <ModalHeader sx={finalHeading ? styles.modalHeader : undefined}>
                {finalHeading}
              </ModalHeader>
              <ModalCloseButton sx={styles.modalCloseButton} />
              <ModalBody sx={styles.modalBody}>
                <Stack direction="column" rowGap="s" spacing="0">
                  {renderChildren({
                    isBlockElement: true,
                    layout: "column",
                    width: "full",
                  })}
                </Stack>
              </ModalBody>
              <ModalFooter sx={styles.modalFooter}>
                <ButtonGroup layout="column" buttonWidth="full">
                  <Button
                    id={`filter-bar-${id}-see-results`}
                    buttonType="primary"
                    type="submit"
                    onClick={onSubmit ? onSubmitAndClose : finalOnClose}
                    ref={showResultsButtonRef}
                  >
                    {`Show ${totalResults ?? ""} results`}
                  </Button>
                  {onClear && (
                    <Button
                      id={`filter-bar-${id}-clear`}
                      buttonType="text"
                      type="reset"
                      onClick={onClearAndFocus}
                      textAlign="center"
                    >
                      Clear all filters
                    </Button>
                  )}
                </ButtonGroup>
              </ModalFooter>
            </ModalContent>
          </Modal>
        </Box>
      );
    }
  )
);

export function useFilterBarPopup(): any {
  const { isOpen, onClose, onOpen } = useDisclosure();
  const InternalFilterBarPopup = chakra((props) => {
    return (
      <FilterBarPopup
        isOpen={isOpen}
        onClose={onClose}
        onOpen={onOpen}
        {...props}
      />
    );
  });

  return { isOpen, onClose, onOpen, FilterBarPopup: InternalFilterBarPopup };
}

export default FilterBarPopup;
