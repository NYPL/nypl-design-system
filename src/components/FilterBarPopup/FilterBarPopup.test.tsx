import { axe } from "jest-axe";
import { render, screen, waitFor } from "@testing-library/react";
import renderer from "react-test-renderer";
import userEvent from "@testing-library/user-event";
import FilterBarPopup from "./FilterBarPopup";
import MultiSelect from "../MultiSelect/MultiSelect";
import MultiSelectGroup from "../MultiSelectGroup/MultiSelectGroup";
import useMultiSelect from "../../hooks/useMultiSelect";

const multiSelectItems = [
  {
    id: "colors",
    name: "Colors",
    items: [
      { id: "red", name: "Red" },
      { id: "blue", name: "Blue" },
      { id: "yellow", name: "Yellow" },
    ],
  },
  {
    id: "pets",
    name: "Pets",
    items: [
      { id: "cat", name: "Cat" },
      {
        id: "dog",
        name: "Dog",
        children: [
          { id: "corgy", name: "Corgy" },
          { id: "german-sheperd", name: "German Sheperd" },
          { id: "afghan-hound", name: "Afghan Hound" },
        ],
      },
      { id: "rat", name: "Rat" },
    ],
  },
  {
    id: "tools",
    name: "Tools",
    items: [
      { id: "hammer", name: "Hammer" },
      {
        id: "skrewdriver",
        name: "Skrewdriver",
        children: [
          { id: "slottet", name: "Slotted Skrewdriver" },
          { id: "phillips", name: "Phillips Skrewdriver" },
          { id: "allen", name: "Allen Skrewdriver" },
        ],
      },
      { id: "whisk", name: "Whisk" },
    ],
  },
];

const defaultItemsVisible = 5;

interface FilterBarTestComponentProps {
  id?: string;
  onClearFilters?: () => void;
  onSubmit?: () => void;
}
const FilterBarTestComponent = ({
  id = "filterbar-id",
  onClearFilters,
  onSubmit,
}: FilterBarTestComponentProps) => {
  const { onChange, onMixedStateChange, selectedItems, onClear, onClearAll } =
    useMultiSelect();
  return (
    <FilterBarPopup
      id={id}
      heading="FilterBar"
      {...(onClearFilters && { onClear: onClearAll })}
      {...(onSubmit && { onSubmit: onSubmit })}
      selectedItems={[selectedItems]}
      renderChildren={({ layout, width }) => {
        return (
          <MultiSelectGroup
            id="multiselect-group"
            key="multiselect-group-key"
            labelText="MultiSelect Group"
            layout={layout}
            multiSelectWidth={width}
            renderMultiSelect={() => {
              return multiSelectItems.map((multiSelect) => (
                <MultiSelect
                  key={multiSelect.id}
                  id={multiSelect.id}
                  items={multiSelect.items}
                  selectedItems={selectedItems}
                  isDefaultOpen={false}
                  isSearchable={false}
                  isBlockElement={false}
                  buttonText={multiSelect.name}
                  defaultItemsVisible={defaultItemsVisible}
                  onChange={(e) => {
                    onChange(e.target.id, multiSelect.id);
                  }}
                  onMixedStateChange={(e) => {
                    onMixedStateChange(
                      e.target.id,
                      multiSelect.id,
                      multiSelect.items
                    );
                  }}
                  onClear={() => {
                    onClear(multiSelect.id);
                  }}
                />
              ));
            }}
          />
        );
      }}
    />
  );
};
const MultiSelectTestGroup = ({ multiSelectItems }) => {
  return (
    <MultiSelectGroup
      id="MultiSelectGroup"
      labelText="MultiSelectGroup example"
      showLabel={true}
      multiSelectWidth="full"
      renderMultiSelect={() => {
        return (
          multiSelectItems &&
          multiSelectItems.map((multiSelectItem) => (
            <MultiSelect
              key={multiSelectItem.id}
              id={multiSelectItem.id}
              items={multiSelectItem.items}
              isDefaultOpen={false}
              isSearchable={false}
              isBlockElement={false}
              selectedItems={{}}
              buttonText="MultiSelect"
              defaultItemsVisible={defaultItemsVisible}
              onChange={() => null}
              onMixedStateChange={() => null}
              onClear={() => "onClear"}
            />
          ))
        );
      }}
    />
  );
};
describe("FilterBarPopup Accessibility", () => {
  beforeAll(() => {
    window.resizeTo = function resizeTo(width, height) {
      Object.assign(this, {
        innerWidth: width,
        innerHeight: height,
        outerWidth: width,
        outerHeight: height,
      }).dispatchEvent(new this.Event("resize"));
    };
  });

  it("should have no axe violations on mobile", async () => {
    window.resizeTo(300, 300);
    const { container } = render(<FilterBarTestComponent />);
    // Make sure it renders mobile components
    expect(
      screen.getByRole("button", { name: /show filters/i })
    ).toBeInTheDocument();

    expect(await axe(container)).toHaveNoViolations();
  });
});

describe("FilterBarPopup", () => {
  beforeAll(() => {
    window.resizeTo = function resizeTo(width, height) {
      Object.assign(this, {
        innerWidth: width,
        innerHeight: height,
        outerWidth: width,
        outerHeight: height,
      }).dispatchEvent(new this.Event("resize"));
    };
    window.resizeTo(300, 300);
  });
  it("should render the `Show filters` button when window size is mobile", () => {
    render(<FilterBarTestComponent id="filter-bar-test-5" />);
    expect(
      screen.getByRole("button", { name: /show filters/i })
    ).toBeInTheDocument();
  });
  it("should open the filter modal whene `Show filters` button is clicked", async () => {
    render(<FilterBarTestComponent id="filter-bar-test-5" />);
    expect(
      screen.getByRole("button", { name: /show filters/i })
    ).toBeInTheDocument();
    // Open the modal
    userEvent.click(screen.getByRole("button", { name: /show filters/i }));
    // Check that all the multiselects are present
    expect(screen.getByRole("button", { name: /pets/i })).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /colors/i })).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /tools/i })).toBeInTheDocument();
    // Check for the Close button to be present
    expect(screen.getByRole("button", { name: /close/i })).toBeInTheDocument();
    // Close the modal
    userEvent.click(screen.getByRole("button", { name: /close/i }));
    // Wait for the modal to close before making the check
    await waitFor(() => {
      expect(
        screen.getByRole("button", { name: /show filters/i })
      ).toBeInTheDocument();
    });
  });
  it("should clear the selectedItems when 'Clear all filters' button is clicked", async () => {
    render(
      <FilterBarTestComponent
        onClearFilters={jest.fn()}
        id="filter-bar-test-2"
      />
    );
    // Open the modal
    userEvent.click(screen.getByRole("button", { name: /show filters/i }));
    // Open menu
    userEvent.click(screen.getByRole("button", { name: /colors/i }));
    // Open menu
    userEvent.click(screen.getByRole("button", { name: /pets/i }));
    // Open menu
    userEvent.click(screen.getByRole("button", { name: /tools/i }));
    // Check item
    userEvent.click(screen.getByRole("checkbox", { name: /blue/i }));
    // Check item
    userEvent.click(screen.getByText("Dog"));
    // Check item
    userEvent.click(screen.getByRole("checkbox", { name: /hammer/i }));
    expect(screen.getAllByRole("button", { name: /selected/i }).length).toBe(6);
    // click Clear All button
    userEvent.click(screen.getByRole("button", { name: /clear all filters/i }));
    expect(
      Object.keys(screen.findAllByRole("button", { name: /selected/i })).length
    ).toBe(0);
  });
  it("should call onSubmit function when 'Show results' button is clicked", () => {
    const mockonSubmit = jest.fn();
    const { rerender } = render(
      <FilterBarTestComponent onSubmit={mockonSubmit} id="filter-bar-test-4" />
    );
    rerender(
      <FilterBarTestComponent onSubmit={mockonSubmit} id="filter-bar-test-4" />
    );
    // Open the modal
    userEvent.click(screen.getByRole("button", { name: /show filters/i }));
    // click Submit All button
    userEvent.click(screen.getByRole("button", { name: /show results/i }));
    expect(mockonSubmit).toBeCalledTimes(1);
  });
  it("renders the UI snapshots correctly", () => {
    const defaultFilterBarNoButtons = renderer
      .create(
        <FilterBarPopup
          id="defaultFilterBarNoButtons"
          heading="FilterBar"
          selectedItems={[]}
          renderChildren={() => {
            return <MultiSelectTestGroup multiSelectItems={multiSelectItems} />;
          }}
        />
      )
      .toJSON();
    const filterBarWithClearAllButton = renderer
      .create(
        <FilterBarPopup
          id="filterBarWithClearAllButton"
          heading="FilterBar"
          onClear={() => null}
          selectedItems={[]}
          renderChildren={() => {
            return <MultiSelectTestGroup multiSelectItems={multiSelectItems} />;
          }}
        />
      )
      .toJSON();
    const filterBarWithSubmitAllButton = renderer
      .create(
        <FilterBarPopup
          id="filterBarWithSubmitAllButton"
          heading="FilterBar"
          onSubmit={() => null}
          selectedItems={[]}
          renderChildren={() => {
            return <MultiSelectTestGroup multiSelectItems={multiSelectItems} />;
          }}
        />
      )
      .toJSON();
    const filterBarWithButtons = renderer
      .create(
        <FilterBarPopup
          id="filterBarWithButtons"
          heading="FilterBar"
          onClear={() => null}
          onSubmit={() => null}
          selectedItems={[]}
          renderChildren={() => {
            return <MultiSelectTestGroup multiSelectItems={multiSelectItems} />;
          }}
        />
      )
      .toJSON();

    expect(defaultFilterBarNoButtons).toMatchSnapshot();
    expect(filterBarWithClearAllButton).toMatchSnapshot();
    expect(filterBarWithSubmitAllButton).toMatchSnapshot();
    expect(filterBarWithButtons).toMatchSnapshot();
  });
});
