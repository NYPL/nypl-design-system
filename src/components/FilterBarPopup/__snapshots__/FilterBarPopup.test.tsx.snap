// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBarPopup renders the UI snapshots correctly 1`] = `
<div
  className="css-1xdhyk6"
  id="filter-bar-defaultFilterBarNoButtons"
>
  <button
    className="chakra-button css-1xdhyk6"
    id="filter-bar-defaultFilterBarNoButtons-show-filters"
    onClick={[Function]}
    type="button"
  >
    Show filters
  </button>
</div>
`;

exports[`FilterBarPopup renders the UI snapshots correctly 2`] = `
<div
  className="css-1xdhyk6"
  id="filter-bar-filterBarWithClearAllButton"
>
  <button
    className="chakra-button css-1xdhyk6"
    id="filter-bar-filterBarWithClearAllButton-show-filters"
    onClick={[Function]}
    type="button"
  >
    Show filters
  </button>
</div>
`;

exports[`FilterBarPopup renders the UI snapshots correctly 3`] = `
<div
  className="css-1xdhyk6"
  id="filter-bar-filterBarWithSubmitAllButton"
>
  <button
    className="chakra-button css-1xdhyk6"
    id="filter-bar-filterBarWithSubmitAllButton-show-filters"
    onClick={[Function]}
    type="button"
  >
    Show filters
  </button>
</div>
`;

exports[`FilterBarPopup renders the UI snapshots correctly 4`] = `
<div
  className="css-1xdhyk6"
  id="filter-bar-filterBarWithButtons"
>
  <button
    className="chakra-button css-1xdhyk6"
    id="filter-bar-filterBarWithButtons-show-filters"
    onClick={[Function]}
    type="button"
  >
    Show filters
  </button>
</div>
`;
